frontend/src/components/widgets/DataFrame/hooks/useDataEditor.ts

```diff
@@ -0,0 +1,254 @@
+ * @param applyEdits - Callback that needs to be called on all edits. This will also trigger a rerun
+ *                     and send widget state to the backend.
+ *
+ * @returns Glide-data-grid compatible functions for editing capabilities.
+ */
+function useDataEditor(
+  columns: BaseColumn[],
+  fixedNumRows: boolean,
+  editingState: React.MutableRefObject<EditingState>,
+  getCellContent: ([col, row]: readonly [number, number]) => GridCell,
+  getOriginalIndex: (index: number) => number,
+  refreshCells: (
+    cells: {
+      cell: [number, number]
+    }[]
+  ) => void,
+  applyEdits: (clearSelection?: boolean, triggerRerun?: boolean) => void
+): DataEditorReturn {
+  const onCellEdited = React.useCallback(
+    (
+      [col, row]: readonly [number, number],
+      updatedCell: EditableGridCell
+    ): void => {
+      const column = columns[col]
+
+      const originalCol = column.indexNumber
+
+      // We need to apply two different mappings here. One for the case that
+      // the user has sorted a column, and another one from the editing state
+      // to get the correct row ID when the user has deleted rows.
+      const originalRow = editingState.current.getOriginalRowIndex(
+        getOriginalIndex(row)
+      )
+      const currentCell = getCellContent([col, row])
+      const currentValue = column.getCellValue(currentCell)
+      const newValue = column.getCellValue(updatedCell)
+      if (!isErrorCell(currentCell) && newValue === currentValue) {
+        // No editing is required since the values did not change
+        return
+      }
+
+      const newCell = column.getCell(newValue)
+
+      editingState.current.setCell(originalCol, originalRow, {
+        ...newCell,
+        lastUpdated: performance.now(),
+      })
+
+      applyEdits()
+    },
+    [columns, editingState, getOriginalIndex, getCellContent, applyEdits]
+  )
+
+  const onRowAppended = React.useCallback(() => {
+    const newRow: Map<number, GridCell> = new Map()
+    columns.forEach(column => {
+      newRow.set(column.indexNumber, column.getCell(undefined))
+    })
+    editingState.current.addRow(newRow)
+    applyEdits(false, false)
+  }, [columns, editingState])
+
+  const onDelete = React.useCallback(
+    (selection: GridSelection): GridSelection | boolean => {
+      if (selection.rows.length > 0) {
+        // User has selected one or more rows
+        if (fixedNumRows) {
+          // Deleting rows is not supported
+          return true
+        }
+
+        const rowsToDelete = selection.rows.toArray().map(row => {
+          return editingState.current.getOriginalRowIndex(
+            getOriginalIndex(row)
+          )
+        })
+        // We need to delete all rows at once, so that the indexes work correct
+        editingState.current.deleteRows(rowsToDelete)
+        applyEdits(true)
+        return false
+      }
+      if (selection.current?.range) {
```

Comment:
Should this code not cover lines 115 - 131? It seems like `selection.current?.range` should return everything that's selected? Or maybe this is because of added rows?
Personally, I find this method kind of confusing.

Reply:
This is the glide-data-grid selection format, which is slightly confusing: But `selection.rows.length` refers to selected rows (e.g. for deleting rows), and `selection.current?.range` refers to selected cells (e.g. if the user uses drag and drop for selecting multiple cells)